From 7e7ad0680096974cc3527fa00c1dc4b89ec2505e Mon Sep 17 00:00:00 2001
From: Ikey Doherty <michael.i.doherty@intel.com>
Date: Wed, 24 Feb 2016 15:16:25 +0000
Subject: [PATCH] math: Disable broken test

Signed-off-by: Ikey Doherty <michael.i.doherty@intel.com>
---
 math/Makefile | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/math/Makefile b/math/Makefile
index 7d573a0..222ee6b 100644
--- a/math/Makefile
+++ b/math/Makefile
@@ -114,7 +114,6 @@ tests = test-matherr test-fenv atest-exp atest-sincos atest-exp2 basic-test \
 	test-nearbyint-except-2 test-signgam-uchar test-signgam-uchar-init \
 	test-signgam-uint test-signgam-uint-init test-signgam-ullong \
 	test-signgam-ullong-init test-nan-overflow test-nan-payload \
-	test-math-isinff \
 	$(tests-static)
 tests-static = test-fpucw-static test-fpucw-ieee-static \
 	       test-signgam-uchar-static test-signgam-uchar-init-static \
@@ -221,8 +220,6 @@ CFLAGS-test-signgam-ullong-init.c = -std=c99
 CFLAGS-test-signgam-ullong-static.c = -std=c99
 CFLAGS-test-signgam-ullong-init-static.c = -std=c99
 
-CFLAGS-test-math-isinff.cc = -std=gnu++11
-
 # The -lieee module sets the _LIB_VERSION_ switch to IEEE mode
 # for error handling in the -lm functions.
 install-lib += libieee.a
-- 
2.7.0

