From e742928c1592b43db6809db4f39e67be151cdd27 Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.com>
Date: Thu, 18 Feb 2016 18:31:04 -0200
Subject: [PATCH 03/20] Regenerated configure scripts.

---
 ChangeLog | 2 ++
 configure | 2 +-
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/ChangeLog b/ChangeLog
index ae34c6f..f5ae211 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,5 +1,7 @@
 2016-02-18  Adhemerval Zanella  <adhemerval.zanella@linaro.org>
 
+	* configure: Regenerated.
+
 	* po/libc.pot: Regenerated.
 
 	* po/be.po: Updated translation.
diff --git a/configure b/configure
index 06ea87e..8fe5937 100755
--- a/configure
+++ b/configure
@@ -3864,7 +3864,7 @@ if test -z "$enable_hacker_mode" && test x"$libc_config_ok" != xyes; then
 *** The GNU C library is currently unavailable for this platform.
 *** If you are interested in seeing glibc on this platform visit
 *** the \"How to submit a new port\" in the wiki:
-***   https://sourceware.org/glibc/wiki/HomePage#Development
+***   https://sourceware.org/glibc/wiki/#Development
 *** and join the community!" "$LINENO" 5
     ;;
   esac
-- 
2.8.0

