From 0a321a461c726096585c22f735266b31328232ec Mon Sep 17 00:00:00 2001
From: Carlos O'Donell <carlos@systemhalted.org>
Date: Fri, 26 Feb 2016 02:09:11 -0500
Subject: [PATCH 08/20] Update NEWS.

---
 NEWS | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/NEWS b/NEWS
index ddc2066..509e336 100644
--- a/NEWS
+++ b/NEWS
@@ -5,6 +5,17 @@ See the end for copying conditions.
 Please send GNU C library bug reports via <http://sourceware.org/bugzilla/>
 using `glibc' in the "product" field.
 
+Version 2.23.1
+
+Security related changes:
+
+  [Add security related changes here]
+
+The following bugs are resolved with this release:
+
+  [19679] gcc-4.9.3 C++ exception handling broken due to unaligned stack
+
+
 Version 2.23
 
 * Unicode 8.0.0 Support: Character encoding, character type info, and
-- 
2.8.0

