diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/doasin-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/doasin-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/doasin-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/doasin-fma3.c	2017-07-31 00:33:51.048693532 +0000
@@ -0,0 +1,4 @@
+#define __doasin __doasin_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/doasin.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/dosincos-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/dosincos-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/dosincos-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/dosincos-fma3.c	2017-07-31 00:31:56.670725975 +0000
@@ -0,0 +1,6 @@
+#define __docos __docos_fma3
+#define __dubcos __dubcos_fma3
+#define __dubsin __dubsin_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/dosincos.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_asin.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_asin.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_asin.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_asin.c	2017-07-31 00:43:19.127130397 +0000
@@ -4,18 +4,18 @@
 
 extern double __ieee754_acos_sse2 (double);
 extern double __ieee754_asin_sse2 (double);
-extern double __ieee754_acos_fma4 (double);
-extern double __ieee754_asin_fma4 (double);
+extern double __ieee754_acos_fma3 (double);
+extern double __ieee754_asin_fma3 (double);
 
 libm_ifunc (__ieee754_acos,
-	    HAS_ARCH_FEATURE (FMA4_Usable)
-	    ? __ieee754_acos_fma4
+	    HAS_ARCH_FEATURE (FMA_Usable)
+	    ? __ieee754_acos_fma3
 	    : __ieee754_acos_sse2);
 strong_alias (__ieee754_acos, __acos_finite)
 
 libm_ifunc (__ieee754_asin,
-	    HAS_ARCH_FEATURE (FMA4_Usable)
-	    ? __ieee754_asin_fma4
+	    HAS_ARCH_FEATURE (FMA_Usable)
+	    ? __ieee754_asin_fma3
 	    : __ieee754_asin_sse2);
 strong_alias (__ieee754_asin, __asin_finite)
 
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_asin-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_asin-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_asin-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_asin-fma3.c	2017-07-31 00:32:50.437245921 +0000
@@ -0,0 +1,11 @@
+#define __ieee754_acos __ieee754_acos_fma3
+#define __ieee754_asin __ieee754_asin_fma3
+#define __cos32 __cos32_fma3
+#define __doasin __doasin_fma3
+#define __docos __docos_fma3
+#define __dubcos __dubcos_fma3
+#define __dubsin __dubsin_fma3
+#define __sin32 __sin32_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/e_asin.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_atan2.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_atan2.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_atan2.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_atan2.c	2017-07-31 00:25:40.333717479 +0000
@@ -5,9 +5,10 @@
 extern double __ieee754_atan2_sse2 (double, double);
 extern double __ieee754_atan2_avx (double, double);
 extern double __ieee754_atan2_fma4 (double, double);
+extern double __ieee754_atan2_fma3 (double, double);
 
 libm_ifunc (__ieee754_atan2,
-	    HAS_ARCH_FEATURE (FMA4_Usable) ? __ieee754_atan2_fma4
+	    HAS_ARCH_FEATURE (FMA_Usable) ? __ieee754_atan2_fma3
 	    : (HAS_ARCH_FEATURE (AVX_Usable)
 	       ? __ieee754_atan2_avx : __ieee754_atan2_sse2));
 strong_alias (__ieee754_atan2, __atan2_finite)
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_atan2-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_atan2-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_atan2-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_atan2-fma3.c	2017-07-31 00:25:40.333717479 +0000
@@ -0,0 +1,10 @@
+#define __ieee754_atan2 __ieee754_atan2_fma3
+#define __add __add_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __dvd __dvd_fma3
+#define __mpatan2 __mpatan2_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/e_atan2.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_exp.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_exp.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_exp.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_exp.c	2017-07-31 00:25:40.333717479 +0000
@@ -5,9 +5,10 @@
 extern double __ieee754_exp_sse2 (double);
 extern double __ieee754_exp_avx (double);
 extern double __ieee754_exp_fma4 (double);
+extern double __ieee754_exp_fma3 (double);
 
 libm_ifunc (__ieee754_exp,
-	    HAS_ARCH_FEATURE (FMA4_Usable) ? __ieee754_exp_fma4
+	    HAS_ARCH_FEATURE (FMA_Usable) ? __ieee754_exp_fma3
 	    : (HAS_ARCH_FEATURE (AVX_Usable)
 	       ? __ieee754_exp_avx : __ieee754_exp_sse2));
 strong_alias (__ieee754_exp, __exp_finite)
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_exp-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_exp-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_exp-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_exp-fma3.c	2017-07-31 00:25:40.333717479 +0000
@@ -0,0 +1,6 @@
+#define __ieee754_exp __ieee754_exp_fma3
+#define __exp1 __exp1_fma3
+#define __slowexp __slowexp_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/e_exp.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_log.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_log.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_log.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_log.c	2017-07-31 00:25:40.333717479 +0000
@@ -5,9 +5,10 @@
 extern double __ieee754_log_sse2 (double);
 extern double __ieee754_log_avx (double);
 extern double __ieee754_log_fma4 (double);
+extern double __ieee754_log_fma3 (double);
 
 libm_ifunc (__ieee754_log,
-	    HAS_ARCH_FEATURE (FMA4_Usable) ? __ieee754_log_fma4
+	    HAS_ARCH_FEATURE (FMA_Usable) ? __ieee754_log_fma3
 	    : (HAS_ARCH_FEATURE (AVX_Usable)
 	       ? __ieee754_log_avx : __ieee754_log_sse2));
 strong_alias (__ieee754_log, __log_finite)
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_log-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_log-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_log-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_log-fma3.c	2017-07-31 00:25:40.333717479 +0000
@@ -0,0 +1,8 @@
+#define __ieee754_log __ieee754_log_fma3
+#define __mplog __mplog_fma3
+#define __add __add_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __sub __sub_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/e_log.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_pow.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_pow.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_pow.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_pow.c	2017-07-31 00:25:40.333717479 +0000
@@ -3,11 +3,12 @@
 #include <math_private.h>
 
 extern double __ieee754_pow_sse2 (double, double);
+extern double __ieee754_pow_fma3 (double, double);
 extern double __ieee754_pow_fma4 (double, double);
 
 libm_ifunc (__ieee754_pow,
-	    HAS_ARCH_FEATURE (FMA4_Usable)
-	    ? __ieee754_pow_fma4
+	    HAS_ARCH_FEATURE (FMA_Usable)
+	    ? __ieee754_pow_fma3
 	    : __ieee754_pow_sse2);
 strong_alias (__ieee754_pow, __pow_finite)
 
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_pow-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_pow-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/e_pow-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/e_pow-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,6 @@
+#define __ieee754_pow __ieee754_pow_fma3
+#define __exp1 __exp1_fma3
+#define __slowpow __slowpow_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/e_pow.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/halfulp-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/halfulp-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/halfulp-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/halfulp-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,4 @@
+#define __halfulp __halfulp_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/halfulp.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/Makefile glibc-2.25/sysdeps/x86_64/fpu/multiarch/Makefile
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/Makefile	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/Makefile	2017-07-31 00:45:37.019715580 +0000
@@ -7,7 +7,15 @@ libm-sysdep_routines += e_exp-fma4 e_log
 			mplog-fma4 mpa-fma4 slowexp-fma4 slowpow-fma4 \
 			sincos32-fma4 doasin-fma4 dosincos-fma4 \
 			halfulp-fma4 mpexp-fma4 \
-			mpatan2-fma4 mpatan-fma4 mpsqrt-fma4 mptan-fma4
+			mpatan2-fma4 mpatan-fma4 mpsqrt-fma4 mptan-fma4 \
+			e_exp-fma3 e_log-fma3 e_pow-fma3 s_atan-fma3 \
+			e_asin-fma3 e_atan2-fma3 s_sin-fma3 s_tan-fma3 \
+			mplog-fma3 mpa-fma3 slowexp-fma3 slowpow-fma3 \
+			sincos32-fma3 doasin-fma3 dosincos-fma3 \
+			halfulp-fma3 mpexp-fma3 \
+			mpatan2-fma3 mpatan-fma3 mpsqrt-fma3 mptan-fma3 
+			
+			
 
 CFLAGS-doasin-fma4.c = -mfma4
 CFLAGS-dosincos-fma4.c = -mfma4
@@ -31,6 +39,28 @@ CFLAGS-slowpow-fma4.c = -mfma4
 CFLAGS-s_sin-fma4.c = -mfma4
 CFLAGS-s_tan-fma4.c = -mfma4
 
+CFLAGS-doasin-fma3.c = -mfma -mavx2
+CFLAGS-dosincos-fma3.c = -mfma -mavx2
+CFLAGS-e_asin-fma3.c = -mfma -mavx2
+CFLAGS-e_atan2-fma3.c = -mfma -mavx2
+CFLAGS-e_exp-fma3.c = -mfma -mavx2
+CFLAGS-e_log-fma3.c = -mfma -mavx2
+CFLAGS-e_pow-fma3.c = -mfma $(config-cflags-nofma)  -mavx2
+CFLAGS-halfulp-fma3.c = -mfma -mavx2
+CFLAGS-mpa-fma3.c = -mfma -mavx2
+CFLAGS-mpatan-fma3.c = -mfma -mavx2
+CFLAGS-mpatan2-fma3.c = -mfma -mavx2
+CFLAGS-mpexp-fma3.c = -mfma -mavx2
+CFLAGS-mplog-fma3.c = -mfma -mavx2
+CFLAGS-mpsqrt-fma3.c = -mfma -mavx2
+CFLAGS-mptan-fma3.c = -mfma -mavx2
+CFLAGS-s_atan-fma3.c = -mfma -mavx2
+CFLAGS-sincos32-fma3.c = -mfma -mavx2
+CFLAGS-slowexp-fma3.c = -mfma -mavx2
+CFLAGS-slowpow-fma3.c = -mfma -mavx2
+CFLAGS-s_sin-fma3.c = -mfma -mavx2
+CFLAGS-s_tan-fma3.c = -mfma -mavx2
+
 libm-sysdep_routines += e_exp-avx e_log-avx s_atan-avx \
 			e_atan2-avx s_sin-avx s_tan-avx \
 			mplog-avx mpa-avx slowexp-avx \
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpa-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpa-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpa-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpa-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,14 @@
+#define __add __add_fma3
+#define __mul __mul_fma3
+#define __sqr __sqr_fma3
+#define __sub __sub_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __dvd __dvd_fma3
+
+#define NO___CPY 1
+#define NO___MP_DBL 1
+#define NO___ACR 1
+#define NO__CONST 1
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/mpa.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpatan2-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpatan2-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpatan2-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpatan2-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,9 @@
+#define __mpatan2 __mpatan2_fma3
+#define __add __add_fma3
+#define __dvd __dvd_fma3
+#define __mpatan __mpatan_fma3
+#define __mpsqrt __mpsqrt_fma3
+#define __mul __mul_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/mpatan2.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpatan-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpatan-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpatan-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpatan-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,10 @@
+#define __mpatan __mpatan_fma3
+#define __add __add_fma3
+#define __dvd __dvd_fma3
+#define __mpsqrt __mpsqrt_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define AVOID_MPATAN_H 1
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/mpatan.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpexp-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpexp-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpexp-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpexp-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,9 @@
+#define __mpexp __mpexp_fma3
+#define __add __add_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __dvd __dvd_fma3
+#define __mul __mul_fma3
+#define AVOID_MPEXP_H 1
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/mpexp.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mplog-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/mplog-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mplog-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/mplog-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,8 @@
+#define __mplog __mplog_fma3
+#define __add __add_fma3
+#define __mpexp __mpexp_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/mplog.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpsqrt-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpsqrt-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mpsqrt-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/mpsqrt-fma3.c	2017-07-31 00:25:40.334717472 +0000
@@ -0,0 +1,8 @@
+#define __mpsqrt __mpsqrt_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define AVOID_MPSQRT_H 1
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/mpsqrt.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mptan-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/mptan-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/mptan-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/mptan-fma3.c	2017-07-31 00:34:44.005202103 +0000
@@ -0,0 +1,7 @@
+#define __mptan __mptan_fma3
+#define __c32 __c32_fma3
+#define __dvd __dvd_fma3
+#define __mpranred __mpranred_fma3
+#define SECTION __attribute__ ((section (".text.fma")))
+
+#include <sysdeps/ieee754/dbl-64/mptan.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_atan.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_atan.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_atan.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_atan.c	2017-07-31 00:43:45.582860229 +0000
@@ -3,9 +3,9 @@
 
 extern double __atan_sse2 (double);
 extern double __atan_avx (double);
-extern double __atan_fma4 (double);
+extern double __atan_fma3 (double);
 
-libm_ifunc (atan, (HAS_ARCH_FEATURE (FMA4_Usable) ? __atan_fma4 :
+libm_ifunc (atan, (HAS_ARCH_FEATURE (FMA_Usable) ? __atan_fma3 :
 		   HAS_ARCH_FEATURE (AVX_Usable)
 		   ? __atan_avx : __atan_sse2));
 
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_atan-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_atan-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_atan-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_atan-fma3.c	2017-07-31 00:36:51.760987757 +0000
@@ -0,0 +1,9 @@
+#define atan __atan_fma3
+#define __add __add_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __mpatan __mpatan_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/s_atan.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/sincos32-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/sincos32-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/sincos32-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/sincos32-fma3.c	2017-07-31 00:36:14.892341994 +0000
@@ -0,0 +1,15 @@
+#define __cos32 __cos32_fma3
+#define __sin32 __sin32_fma3
+#define __c32 __c32_fma3
+#define __mpsin __mpsin_fma3
+#define __mpsin1 __mpsin1_fma3
+#define __mpcos __mpcos_fma3
+#define __mpcos1 __mpcos1_fma3
+#define __mpranred __mpranred_fma3
+#define __add __add_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/sincos32.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/slowexp-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/slowexp-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/slowexp-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/slowexp-fma3.c	2017-07-31 00:25:40.335717466 +0000
@@ -0,0 +1,9 @@
+#define __slowexp __slowexp_fma3
+#define __add __add_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __mpexp __mpexp_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/slowexp.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/slowpow-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/slowpow-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/slowpow-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/slowpow-fma3.c	2017-07-31 00:25:40.335717466 +0000
@@ -0,0 +1,11 @@
+#define __slowpow __slowpow_fma3
+#define __add __add_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __mpexp __mpexp_fma3
+#define __mplog __mplog_fma3
+#define __mul __mul_fma3
+#define __sub __sub_fma3
+#define __halfulp __halfulp_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/slowpow.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_sin.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_sin.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_sin.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_sin.c	2017-07-31 00:43:37.632941470 +0000
@@ -6,15 +6,15 @@ extern double __cos_sse2 (double);
 extern double __sin_sse2 (double);
 extern double __cos_avx (double);
 extern double __sin_avx (double);
-extern double __cos_fma4 (double);
-extern double __sin_fma4 (double);
+extern double __cos_fma3 (double);
+extern double __sin_fma3 (double);
 
-libm_ifunc (__cos, (HAS_ARCH_FEATURE (FMA4_Usable) ? __cos_fma4 :
+libm_ifunc (__cos, (HAS_ARCH_FEATURE (FMA_Usable) ? __cos_fma3 :
 		    HAS_ARCH_FEATURE (AVX_Usable)
 		    ? __cos_avx : __cos_sse2));
 weak_alias (__cos, cos)
 
-libm_ifunc (__sin, (HAS_ARCH_FEATURE (FMA4_Usable) ? __sin_fma4 :
+libm_ifunc (__sin, (HAS_ARCH_FEATURE (FMA_Usable) ? __sin_fma3 :
 		    HAS_ARCH_FEATURE (AVX_Usable)
 		    ? __sin_avx : __sin_sse2));
 weak_alias (__sin, sin)
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_sin-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_sin-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_sin-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_sin-fma3.c	2017-07-31 00:35:48.244596205 +0000
@@ -0,0 +1,11 @@
+#define __cos __cos_fma3
+#define __sin __sin_fma3
+#define __docos __docos_fma3
+#define __dubsin __dubsin_fma3
+#define __mpcos __mpcos_fma3
+#define __mpcos1 __mpcos1_fma3
+#define __mpsin __mpsin_fma3
+#define __mpsin1 __mpsin1_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/s_sin.c>
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_tan.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_tan.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_tan.c	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_tan.c	2017-07-31 00:43:28.453035224 +0000
@@ -3,9 +3,9 @@
 
 extern double __tan_sse2 (double);
 extern double __tan_avx (double);
-extern double __tan_fma4 (double);
+extern double __tan_fma3 (double);
 
-libm_ifunc (tan, (HAS_ARCH_FEATURE (FMA4_Usable) ? __tan_fma4 :
+libm_ifunc (tan, (HAS_ARCH_FEATURE (FMA_Usable) ? __tan_fma3 :
 		  HAS_ARCH_FEATURE (AVX_Usable)
 		  ? __tan_avx : __tan_sse2));
 
diff -purN glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_tan-fma3.c glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_tan-fma3.c
--- glibc-2.25/sysdeps/x86_64/fpu.org/multiarch/s_tan-fma3.c	1970-01-01 00:00:00.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/fpu/multiarch/s_tan-fma3.c	2017-07-31 00:35:27.260795258 +0000
@@ -0,0 +1,8 @@
+#define tan __tan_fma3
+#define __dbl_mp __dbl_mp_fma3
+#define __mpranred __mpranred_fma3
+#define __mptan __mptan_fma3
+#define __sub __sub_fma3
+#define SECTION __attribute__ ((section (".text.fma3")))
+
+#include <sysdeps/ieee754/dbl-64/s_tan.c>
