Subject: make x86_64 glibc build of ldconfig not search libx32 & lib
Index: glibc-2.21/sysdeps/unix/sysv/linux/x86_64/dl-cache.h
===================================================================
--- glibc-2.21.orig/sysdeps/unix/sysv/linux/x86_64/dl-cache.h	2015-02-06 06:40:18.000000000 +0000
+++ glibc-2.21/sysdeps/unix/sysv/linux/x86_64/dl-cache.h	2015-03-03 14:56:33.106998797 +0000
@@ -21,31 +21,4 @@
 #define _dl_cache_check_flags(flags)			\
   ((flags) == _DL_CACHE_DEFAULT_ID)
 
-#define add_system_dir(dir) \
-  do								\
-    {								\
-      size_t len = strlen (dir);				\
-      char path[len + 4];					\
-      memcpy (path, dir, len + 1);				\
-      if (len >= 6 && ! memcmp (path + len - 6, "/lib64", 6))	\
-	{							\
-	  len -= 2;						\
-	  path[len] = '\0';					\
-	}							\
-      else if (len >= 7						\
-	       && ! memcmp (path + len - 7, "/libx32", 7))	\
-	{							\
-	  len -= 3;						\
-	  path[len] = '\0';					\
-	}							\
-      add_dir (path);						\
-      if (len >= 4 && ! memcmp (path + len - 4, "/lib", 4))	\
-	{							\
-	  memcpy (path + len, "64", 3);				\
-	  add_dir (path);					\
-	  memcpy (path + len, "x32", 4);				\
-	  add_dir (path);					\
-	}							\
-    } while (0)
-
 #include <sysdeps/generic/dl-cache.h>
