diff -purN glibc-2.25/sysdeps/x86_64.org/multiarch/memcpy.S glibc-2.25/sysdeps/x86_64/multiarch/memcpy.S
--- glibc-2.25/sysdeps/x86_64.org/multiarch/memcpy.S	2017-02-05 15:28:43.000000000 +0000
+++ glibc-2.25/sysdeps/x86_64/multiarch/memcpy.S	2017-04-01 17:26:50.653481909 +0000
@@ -33,7 +33,7 @@ ENTRY(__new_memcpy)
 	HAS_ARCH_FEATURE (Prefer_ERMS)
 	jnz	2f
 	HAS_ARCH_FEATURE (AVX512F_Usable)
-	jz	1f
+	jmp	1f
 	lea	__memcpy_avx512_no_vzeroupper(%rip), %RAX_LP
 	HAS_ARCH_FEATURE (Prefer_No_VZEROUPPER)
 	jnz	2f
